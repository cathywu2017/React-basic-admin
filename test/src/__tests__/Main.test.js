import React from 'react';
import {
  cleanup,
  render,
  waitForElement,
  fireEvent,
} from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Main from '../Main';

afterEach(cleanup);

it('renders table data correctly', async () => {
  const fakeUser = [
    {
      id: 1,
      username: 'user1',
      created_at: '2019/11/20 8:53:26',
      enable: 0,
      locked: 0,
    },
  ];

  const pagination = {
    first_result: 0,
    max_results: 20,
    total: 1,
  };

  jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve({
      ret: fakeUser,
      pagination,
    }),
  }));

  const { getByText, container } = render(
    <Main />,
  );

  const totalLen = await waitForElement(() => pagination.total);

  expect(totalLen).toBe(1);

  act(() => {
    fireEvent.click(getByText('新增'));
    fireEvent.click(getByText('編輯'));
    fireEvent.click(getByText('刪除'));
    fireEvent.click(getByText('搜尋'));
  });

  act(() => {
    fireEvent.click(document.querySelector('.modals'));
  });

  expect(container).toMatchSnapshot();

  global.fetch.mockRestore();
});

it('changePage test', async () => {
  const fakeUser = [
    {
      id: 1,
      username: 'user1',
      created_at: '2019/11/20 8:53:26',
      enable: 1,
      locked: 1,
    },
    {
      id: 2,
      username: 'user2',
      created_at: '2019/11/20 8:53:26',
      enable: 1,
      locked: 0,
    },
  ];

  const pagination = {
    first_result: 0,
    max_results: 1,
    total: 2,
  };

  const updata = jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve({
      ret: fakeUser,
      pagination,
    }),
  }));

  const { getAllByText } = render(<Main />);

  const totalLen = await waitForElement(() => pagination.total);

  expect(totalLen).toBe(2);

  await act(async () => {
    fireEvent.click(getAllByText('»')[0]);
  });

  expect(updata).toHaveBeenLastCalledWith('/api/users?first_result=1', { method: 'GET' });

  global.fetch.mockRestore();
});

it('afer search changePage test', async () => {
  const fakeUser = [
    {
      id: 1,
      username: 'user1',
      created_at: '2019/11/20 8:53:26',
      enable: 1,
      locked: 1,
    },
    {
      id: 2,
      username: 'user2',
      created_at: '2019/11/20 8:53:26',
      enable: 1,
      locked: 0,
    },
  ];

  const pagination = {
    first_result: 0,
    max_results: 1,
    total: 2,
  };

  const updata = jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve({
      ret: fakeUser,
      pagination,
    }),
  }));

  const { getAllByText, getByText, getByPlaceholderText } = render(<Main />);

  expect(await waitForElement(() => pagination.total)).toBe(2);

  fireEvent.click(getByText('搜尋'));

  expect(document.querySelector('.modal')).toBeTruthy();

  fireEvent.change(getByPlaceholderText('請輸入帳號'), { target: { value: 'user1' } });

  fireEvent.click(getAllByText('搜尋')[2]);

  await act(async () => {
    expect(updata).toHaveBeenLastCalledWith('/api/users?first_result=0&username=user1', { method: 'GET' });
  });

  await act(async () => {
    fireEvent.click(getAllByText('»')[0]);
  });

  global.fetch.mockRestore();
});

it('export CSV', async () => {
  const fakeUser = [
    {
      id: 1,
      username: 'user1',
      created_at: '2019/11/20 8:53:26',
      enable: 1,
      locked: 1,
    },
    {
      id: 2,
      username: 'user2',
      created_at: '2019/11/20 8:53:26',
      enable: 0,
      locked: 0,
    },
  ];

  const pagination = {
    first_result: 0,
    max_results: 1,
    total: 2,
  };

  jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve({
      ret: fakeUser,
      pagination,
    }),
  }));

  const { getByText } = render(<Main />);

  await waitForElement(() => pagination.total);

  global.URL.createObjectURL = jest.fn();

  fireEvent.click(getByText('匯出'));

  global.fetch.mockRestore();
});
