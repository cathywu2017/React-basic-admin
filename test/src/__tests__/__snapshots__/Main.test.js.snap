// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders table data correctly 1`] = `
<div>
  <div
    class="pusher"
  >
    <div
      class="ui basic segment"
    >
      <div
        class="ui middle aligned grid"
      >
        <div
          class="left floated twelve wide column"
        >
          <div>
            <div
              aria-label="Pagination Navigation"
              class="ui pagination menu"
              role="navigation"
            >
              <a
                aria-current="false"
                aria-disabled="true"
                class="disabled item"
                tabindex="-1"
                type="firstItem"
                value="1"
              >
                «
              </a>
              <a
                aria-current="false"
                aria-disabled="true"
                class="disabled icon item"
                tabindex="-1"
                type="prevItem"
                value="1"
              >
                <i
                  aria-hidden="true"
                  class="chevron left icon"
                />
              </a>
              <a
                aria-current="true"
                aria-disabled="false"
                class="active item"
                tabindex="0"
                type="pageItem"
                value="1"
              >
                1
              </a>
              <a
                aria-current="false"
                aria-disabled="true"
                class="disabled icon item"
                tabindex="-1"
                type="nextItem"
                value="1"
              >
                <i
                  aria-hidden="true"
                  class="chevron right icon"
                />
              </a>
              <a
                aria-current="false"
                aria-disabled="true"
                class="disabled item"
                tabindex="-1"
                type="lastItem"
                value="1"
              >
                »
              </a>
            </div>
          </div>
        </div>
        <div
          class="right aligned right floated four wide column"
        >
          <button
            class="ui blue button"
          >
            新增
          </button>
          <button
            class="ui teal button"
          >
            搜尋
          </button>
          <button
            class="ui green button"
          >
            匯出
          </button>
        </div>
      </div>
      <table
        class="ui celled striped table"
      >
        <thead
          class=""
        >
          <tr
            class="center aligned"
          >
            <th
              class="one wide"
            >
              #
            </th>
            <th
              class=""
            >
              使用者帳號
            </th>
            <th
              class="three wide"
            >
              建立時間
            </th>
            <th
              class="two wide"
            >
              啟用
            </th>
            <th
              class="two wide"
            >
              鎖定
            </th>
            <th
              class="three wide"
            >
              操作
            </th>
          </tr>
        </thead>
        <tbody
          class=""
        >
          <tr
            class=""
          >
            <td
              class="center aligned"
            >
              1
            </td>
            <td
              class=""
            >
              user1
            </td>
            <td
              class=""
            >
              2019/11/20 8:53:26
            </td>
            <td
              class="center aligned"
            >
              否
            </td>
            <td
              class="center aligned"
            >
              否
            </td>
            <td
              class="center aligned"
            >
              <button
                class="ui green basic button"
              >
                編輯
              </button>
              <button
                class="ui red basic button"
              >
                刪除
              </button>
            </td>
          </tr>
        </tbody>
      </table>
      <div>
        <div
          aria-label="Pagination Navigation"
          class="ui pagination menu"
          role="navigation"
        >
          <a
            aria-current="false"
            aria-disabled="true"
            class="disabled item"
            tabindex="-1"
            type="firstItem"
            value="1"
          >
            «
          </a>
          <a
            aria-current="false"
            aria-disabled="true"
            class="disabled icon item"
            tabindex="-1"
            type="prevItem"
            value="1"
          >
            <i
              aria-hidden="true"
              class="chevron left icon"
            />
          </a>
          <a
            aria-current="true"
            aria-disabled="false"
            class="active item"
            tabindex="0"
            type="pageItem"
            value="1"
          >
            1
          </a>
          <a
            aria-current="false"
            aria-disabled="true"
            class="disabled icon item"
            tabindex="-1"
            type="nextItem"
            value="1"
          >
            <i
              aria-hidden="true"
              class="chevron right icon"
            />
          </a>
          <a
            aria-current="false"
            aria-disabled="true"
            class="disabled item"
            tabindex="-1"
            type="lastItem"
            value="1"
          >
            »
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
